# Cars data - non visual analysis

## Importing libraries

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

## Reading file

In [2]:
data = pd.read_excel('data-cars.xlsx')

## Looking into data

In [3]:
data.sample(5)

Vehicle Name  Sedan  Sports Car  SUV  Wagon  \
257        Mercedes-Benz C320 Sport 4dr      1           0    0      0   
406  Volkswagen New Beetle GLS 1.8T 2dr      1           0    0      0   
216            Land Rover Freelander SE      0           0    1      0   
155                  Honda Civic HX 2dr      1           0    0      0   
265              Mercedes-Benz E500 4dr      1           0    0      0   

     Minivan  Pickup  AWD  RWD  Retail Price  Dealer Cost  Engine Size (l)  \
257        0       0    0    1         35920        33456              3.2   
406        0       0    0    0         21055        19638              1.8   
216        0       0    1    0         25995        23969              2.5   
155        0       0    0    0         14170        12996              1.7   
265        0       0    0    1         57270        53382              5.0   

     Cyl   HP City MPG Hwy MPG Weight Wheel Base  Len Width  
257    6  215       19      26   3430        107  178    68  
406    4  150       24      31   2820         99  161    68  
216    6  174       18      21   3577        101  175    71  
155    4  117       36      44   2500        103  175    67  
265    8  302       16      20   3815        112  190    71

## Data shape

In [4]:
print('We have {0} features'.format(data.shape[1]))
print('We have {0} observations'.format(data.shape[0]))

We have 20 features
We have 428 observations


## Adding 'Car Type' column

For easier grouping of values I've created a single 'Car Type' column with car type names from multiple binary columns.

In [5]:
data['Car Type'] = data[['Sedan', 'Sports Car', 'SUV', 'Wagon', 'Minivan', 'Pickup']].idxmax(axis=1)

In [6]:
data.head()

Vehicle Name  Sedan  Sports Car  SUV  Wagon  Minivan  \
0               Acura 3.5 RL 4dr      1           0    0      0        0   
1  Acura 3.5 RL w/Navigation 4dr      1           0    0      0        0   
2                      Acura MDX      0           0    1      0        0   
3   Acura NSX coupe 2dr manual S      0           1    0      0        0   
4           Acura RSX Type S 2dr      1           0    0      0        0   

   Pickup  AWD  RWD  Retail Price     ...      Engine Size (l)  Cyl   HP  \
0       0    0    0         43755     ...                  3.5    6  225   
1       0    0    0         46100     ...                  3.5    6  225   
2       0    1    0         36945     ...                  3.5    6  265   
3       0    0    1         89765     ...                  3.2    6  290   
4       0    0    0         23820     ...                  2.0    4  200   

   City MPG Hwy MPG Weight Wheel Base  Len Width    Car Type  
0        18      24   3880        115  197    72       Sedan  
1        18      24   3893        115  197    72       Sedan  
2        17      23   4451        106  189    77         SUV  
3        17      24   3153        100  174    71  Sports Car  
4        24      31   2778        101  172    68       Sedan  

[5 rows x 21 columns]

## Analysing Retail Price

In [7]:
data.groupby('Car Type')['Retail Price'].agg(['median', 'mean', 'max', 'min']).round(0).sort_values('median')

median     mean     max    min
Car Type                                   
Pickup      22180.0  24941.0   52975  12800
Wagon       25545.0  28841.0   60670  11905
Sedan       25955.0  29814.0  128420  10280
Minivan     27235.0  27796.0   38380  20615
SUV         32062.0  34790.0   76870  17163
Sports Car  40590.0  53387.0  192465  18345

Sports cars are the most expensive with median price 40590.

## Creating new feature 'Car Make'

In [8]:
data['Car Make'] = data['Vehicle Name'].str.split(' ', n=1, expand=True)[0]

## Creating new feature 'Dealer Margin'

In [9]:
data['Dealer Margin'] = data['Retail Price'] - data['Dealer Cost']

## Creating new feature 'Dealer Margin Percentage'

In [10]:
data['Dealer Margin Percentage'] = round(data['Dealer Margin'] / data['Retail Price'] * 100, 1)

In [11]:
data.head()

Vehicle Name  Sedan  Sports Car  SUV  Wagon  Minivan  \
0               Acura 3.5 RL 4dr      1           0    0      0        0   
1  Acura 3.5 RL w/Navigation 4dr      1           0    0      0        0   
2                      Acura MDX      0           0    1      0        0   
3   Acura NSX coupe 2dr manual S      0           1    0      0        0   
4           Acura RSX Type S 2dr      1           0    0      0        0   

   Pickup  AWD  RWD  Retail Price           ...             City MPG  Hwy MPG  \
0       0    0    0         43755           ...                   18       24   
1       0    0    0         46100           ...                   18       24   
2       0    1    0         36945           ...                   17       23   
3       0    0    1         89765           ...                   17       24   
4       0    0    0         23820           ...                   24       31   

   Weight  Wheel Base  Len Width    Car Type Car Make Dealer Margin  \
0    3880         115  197    72       Sedan    Acura          4741   
1    3893         115  197    72       Sedan    Acura          5000   
2    4451         106  189    77         SUV    Acura          3608   
3    3153         100  174    71  Sports Car    Acura          9787   
4    2778         101  172    68       Sedan    Acura          2059   

  Dealer Margin Percentage  
0                     10.8  
1                     10.8  
2                      9.8  
3                     10.9  
4                      8.6  

[5 rows x 24 columns]

## Analysing 'Dealer Margin'

In [12]:
data.groupby(['Car Make'])['Dealer Margin'].agg(['median', 'mean', 'max', 'min']).round(1).sort_values('median', ascending=False)

median    mean    max   min
Car Make                                  
Porsche        9637.0  9902.1  18905  5479
Jaguar         5885.0  5482.0   7769  2640
Lexus          4814.0  5455.5   8345  3595
CMC            4364.0  4364.0   4364  4364
Hummer         4180.0  4180.0   4180  4180
Cadillac       4173.5  4047.5   5654  2260
Audi           3780.0  3977.8   8183  2432
Mercedes-Benz  3647.0  4203.5   8820  1811
Acura          3608.0  4347.7   9787  2059
Land           3473.0  3980.7   6443  2026
Lincoln        3472.0  3739.8   6415  2526
BMW            3285.0  3664.6   6365  2340
Infiniti       2982.0  3190.0   4970  2259
GMC            2968.0  3115.0   5731  1653
Buick          2460.0  2682.9   3793  1829
Mercury        2447.0  2315.2   3149  1478
Jeep           2219.0  1873.7   2245  1157
Subaru         2216.0  2204.3   2942  1546
Oldsmobile     2190.0  2014.3   2670  1183
Volvo          2183.0  2098.0   2637  1434
Toyota         2176.5  2228.5   6779   616
Volkswagen     2099.0  2566.0   5870  1237
Saab           2094.0  2019.7   2469  1591
Chrysler       2009.5  1996.1   3317  1066
Chevrolet      1944.0  2526.2   6342   725
Mazda          1887.0  1706.9   2150   770
Pontiac        1855.0  1996.7   2998  1072
Ford           1833.0  2062.9   4981   668
Chrvsler       1783.0  1783.0   1783  1783
Mini           1712.0  1712.0   1862  1562
Nissan         1707.0  1733.5   3025   535
Honda          1689.0  1921.9   3295  1095
Mitsubishi     1636.0  1588.8   2349   871
Dodge          1533.0  2093.8   7344   821
Isuzu          1530.0  1530.0   1872  1188
Mazda6         1453.0  1453.0   1453  1453
Hyundai        1285.0  1441.2   2703   432
Mazda3         1126.5  1126.5   1278   975
Saturn         1032.5  1166.4   1781   676
Kia             840.0   985.5   2236   405
Scion           655.0   655.0    685   625
Suzuki          198.0   339.0   1392   153

In [13]:
data.groupby(['Car Make'])['Dealer Margin Percentage'].agg(['median', 'mean', 'max', 'min']).round(1).sort_values('median', ascending=False)

median  mean   max   min
Car Make                               
Porsche          12.6  12.3  14.2   9.8
CMC              12.2  12.2  12.2  12.2
Lexus            11.8  12.2  12.9  11.1
GMC              10.0  10.7  12.4   9.3
Toyota            9.8   9.4  12.4   5.7
Acura             9.8   9.8  10.9   8.6
Audi              9.4   9.3   9.8   6.4
Oldsmobile        9.3   8.3   9.3   6.3
Mini              9.2   9.2   9.3   9.2
Infiniti          9.0   8.8   9.5   7.6
Subaru            8.9   8.6   9.3   7.7
Jaguar            8.9   8.9   8.9   8.8
Land              8.8   8.5   8.9   7.8
BMW               8.4   8.4   8.7   8.2
Cadillac          8.4   8.0   8.6   7.3
Honda             8.4   8.8   9.9   6.3
Hummer            8.4   8.4   8.4   8.4
Buick             8.3   8.7   9.3   8.1
Mercury           8.3   8.2   9.3   6.0
Chevrolet         8.3   8.8  12.5   5.7
Pontiac           8.3   8.1   9.3   6.3
Lincoln           8.2   8.6  12.2   7.8
Volkswagen        8.1   7.9   9.2   6.6
Ford              8.1   8.0  12.3   4.6
Jeep              8.0   7.5   8.8   5.7
Hyundai           7.7   7.8  10.3   4.1
Mazda6            7.5   7.5   7.5   7.5
Mazda             7.4   7.2   8.4   5.2
Dodge             7.3   7.6  10.6   5.9
Chrysler          7.1   7.2   8.6   5.9
Chrvsler          6.9   6.9   6.9   6.9
Mercedes-Benz     6.9   6.9   8.2   6.8
Mitsubishi        6.9   6.7   7.1   6.0
Mazda3            6.8   6.8   7.4   6.3
Nissan            6.8   6.8   8.9   4.2
Saturn            6.4   6.6   7.6   6.1
Isuzu             5.8   5.8   5.9   5.8
Volvo             5.8   5.8   5.8   5.6
Kia               5.7   5.9   8.6   3.9
Saab              5.3   5.4   6.0   4.4
Scion             4.8   4.8   4.8   4.8
Suzuki            1.2   1.8   5.9   1.2

Porsche dealers have the highest margins in absolute values and percentages and Suzuki dealers have the lowest margins.

In [14]:
data.groupby(['Car Type'])['Dealer Margin'].agg(['median', 'mean', 'min', 'max', 'std']).round(1).sort_values('std', ascending=False)

median    mean   min    max     std
Car Type                                       
Sports Car  3499.0  4913.9  1402  18905  3449.0
SUV         2862.5  3164.9   214   8345  1765.5
Sedan       2069.0  2368.2   153   8820  1433.0
Pickup      2003.5  2324.6   668   4941  1204.5
Wagon       1940.0  2194.9   206   4644  1109.7
Minivan     2446.5  2441.0  1215   3317   553.5

In [15]:
data.groupby('Car Type')['Dealer Margin Percentage'].agg(['median', 'mean', 'max', 'min', 'std']).round(1).sort_values('std', ascending=False)

median  mean   max  min  std
Car Type                                
SUV            8.6   8.6  12.9  1.2  2.4
Pickup         9.0   9.0  12.3  4.6  2.3
Sports Car     8.5   8.9  14.2  6.3  2.0
Wagon          8.0   7.4  11.7  1.2  2.0
Sedan          8.1   7.7  12.9  1.2  1.8
Minivan        9.2   8.7  10.8  5.9  1.3

Sports cars have the highest standard deviation in dealers' margin absolute values. Minivan car types have the lowest standard deviation in dealers' margin percentages.

## Changing objects features into integers

In [16]:
# selecting object features
black_list = ['Vehicle Name']
objects = [feature for feature in data.select_dtypes('object') if feature not in black_list]

In [17]:
# replacing string value "*" with integer value -123 ('magic number')
data = data.replace("*",-123)

## Analysing City MPG and Hwy MPG

Grouping Car Types based on City MPG with unselected missing values

In [18]:
data[(data['City MPG'] != -123) & (data['Hwy MPG'] != -123)].groupby('Car Type')['City MPG'].agg(['mean', 'median', 'max', 'min']).round(1).sort_values('median')

mean  median  max  min
Car Type                          
Pickup      16.7      16   24   13
SUV         16.2      16   22   10
Minivan     17.9      18   20   14
Sports Car  18.6      18   26   13
Sedan       21.8      20   60   13
Wagon       21.0      20   31   15

In [19]:
data[(data['City MPG'] != -123) & (data['Hwy MPG'] != -123)].groupby('Car Type')['Hwy MPG'].agg(['mean', 'median', 'max', 'min']).round(1).sort_values('median')

mean  median  max  min
Car Type                          
Pickup      21.2      20   29   17
SUV         20.6      21   27   12
Minivan     24.4      25   27   17
Sports Car  25.7      26   33   19
Wagon       27.8      27   36   19
Sedan       29.4      28   66   19

Sedan is the best fuel efficient car type both in the city and on a highway.

In [20]:
data[(data['City MPG'] != -123) & (data['Hwy MPG'] != -123)].groupby('Car Make')['City MPG'].agg(['mean', 'median', 'max', 'min']).sort_values('median').round(1)

mean  median  max  min
Car Make                             
Hummer         10.0    10.0   10   10
Land           14.0    12.0   18   12
GMC            15.8    15.5   19   13
CMC            16.0    16.0   16   16
Jeep           17.3    16.0   20   16
Isuzu          16.0    16.0   17   15
Mercury        17.6    17.0   20   16
Lincoln        16.8    17.0   20   13
Cadillac       16.5    17.5   18   13
Mercedes-Benz  17.3    17.5   22   13
Infiniti       17.2    17.5   19   15
Porsche        17.4    18.0   20   14
Mitsubishi     18.6    18.0   21   15
Lexus          17.5    18.0   20   13
Jaguar         17.5    18.0   18   16
Acura          19.4    18.0   24   17
Audi           18.5    18.0   23   14
Ford           19.7    18.5   27   14
BMW            18.7    19.0   21   16
Dodge          20.0    19.0   29   15
Chevrolet      19.7    19.0   28   13
Pontiac        21.4    20.0   29   18
Oldsmobile     21.0    20.0   24   19
Nissan         19.7    20.0   28   13
Buick          18.9    20.0   20   15
Volvo          19.8    20.0   22   15
Mazda          20.1    20.0   24   15
Chrysler       19.8    20.0   22   17
Chrvsler       21.0    21.0   21   21
Saab           20.4    21.0   21   19
Subaru         20.3    21.0   22   18
Toyota         24.4    21.5   59   13
Volkswagen     22.5    22.0   38   15
Hyundai        23.0    23.0   29   17
Suzuki         22.1    23.0   25   18
Kia            21.9    24.0   26   16
Mazda6         24.0    24.0   24   24
Saturn         24.4    26.0   26   20
Honda          27.8    26.0   60   17
Mini           26.5    26.5   28   25
Scion          31.5    31.5   32   31

In [21]:
data[(data['City MPG'] != -123) & (data['Hwy MPG'] != -123)].groupby('Car Make')['Hwy MPG'].agg(['mean', 'median', 'max', 'min']).sort_values('median').round(1)

mean  median  max  min
Car Make                             
Hummer         12.0    12.0   12   12
Land           17.7    16.0   21   16
CMC            19.0    19.0   19   19
GMC            19.8    19.5   24   17
Isuzu          20.5    20.5   21   20
Jeep           21.3    21.0   24   19
Lexus          23.5    24.0   29   17
Acura          26.1    24.0   31   23
Infiniti       23.9    24.5   26   19
Mercedes-Benz  23.7    24.5   30   14
Cadillac       23.0    25.0   26   17
Mercury        24.6    25.0   27   21
Mazda          25.4    25.0   29   19
Audi           25.8    25.0   31   20
Lincoln        23.4    25.0   26   18
Ford           26.3    25.5   36   18
Porsche        24.7    26.0   29   18
Nissan         25.4    26.0   35   18
Mitsubishi     25.0    26.0   28   19
Jaguar         25.6    26.0   28   23
Dodge          26.8    26.5   36   21
Volvo          26.7    27.0   29   20
Chrvsler       27.0    27.0   27   27
Chrysler       27.4    27.5   30   23
BMW            27.0    27.5   30   22
Chevrolet      26.9    28.0   37   17
Subaru         27.1    28.0   28   24
Buick          27.8    29.0   30   21
Toyota         30.8    29.0   51   17
Oldsmobile     29.0    29.0   32   26
Saab           28.9    29.0   30   28
Suzuki         27.9    29.5   31   22
Hyundai        29.9    30.0   34   26
Honda          34.0    30.0   66   22
Pontiac        29.7    30.0   36   24
Volkswagen     29.8    30.0   46   20
Kia            28.8    32.0   33   19
Mazda6         32.0    32.0   32   32
Saturn         32.9    35.0   35   26
Mini           35.5    35.5   37   34
Scion          36.5    36.5   38   35

Scion is the best fuel efficient car make and Hummer is the lowest fuel efficient car make.

## Creating new feature with Average MPG

In [22]:
data['Average MPG'] = (data['City MPG'] + data['Hwy MPG']) / 2

## Calculating Retail Price to Average MPG ratio

In [23]:
data['Price To MPG Ratio'] = data['Retail Price'] / data['Average MPG']

## Which car is the most fuel efficient for lowest price?

In [24]:
data[['Vehicle Name', 'Retail Price', 'Price To MPG Ratio']][(data['City MPG'] != -123) & (data['Hwy MPG'] != -123)].sort_values('Price To MPG Ratio')

Vehicle Name  Retail Price  \
387                        Toyota Echo 2dr manual         10760   
388                               Toyota Echo 4dr         11290   
161              Honda Insight 2dr (gas/electric)         19110   
386                          Toyota Echo 2dr auto         11560   
167                      Hyundai Accent 2dr hatch         10539   
208                            Kia Rio 4dr manual         10280   
155                            Honda Civic HX 2dr         14170   
344                               Saturn Ion1 4dr         10995   
352                            Scion xA 4dr hatch         12965   
393               Toyota Prius 4dr (gas/electric)         20510   
65                             Chevrolet Aveo 4dr         11690   
153                            Honda Civic DX 2dr         13270   
168                         Hyundai Accent GL 4dr         11839   
169                   Hyundai Accent GT 2dr hatch         11939   
383                         Toyota Corolla CE 4dr         14085   
207                              Kia Rio 4dr auto         11155   
209                                 Kia Rio Cinco         11905   
311                         Nissan Sentra 1.8 4dr         12740   
66                    Chevrolet Aveo LS 4dr hatch         12585   
156  Honda Civic Hybrid 4dr manual (gas/electric)         20140   
385                          Toyota Corolla S 4dr         15030   
113                             Dodge Neon SE 4dr         13670   
384                         Toyota Corolla LE 4dr         15295   
353                                      Scion xB         14165   
128                             Ford Focus LX 4dr         13730   
212                               Kia Spectra 4dr         12360   
369                          Suzuki Forenza S 4dr         12269   
132                      Ford Focus ZX3 2dr hatch         13270   
157                            Honda Civic LX 4dr         15850   
365                             Suzuki Aeno S 4dr         12884   
..                                            ...           ...   
265                        Mercedes-Benz E500 4dr         57270   
59                             Cadillac Escaladet         52795   
39                                  BMW 745Li 4dr         73195   
19                          Audi A8 L Quattro 4dr         69190   
268                        Mercedes-Benz S430 4dr         74320   
234                      Lincoln Navigator Luxury         52775   
195                    Jaguar XK8 convertible 2dr         74995   
333                   Porsche 911 Targa coupe 2dr         76765   
58                          Cadillac Escalade EXT         52975   
336                             Porsche Cayenne S         56665   
331   Porsche 911 Carrera convertible 2dr (coupe)         79165   
62                   Cadillac XLR convertible 2dr         76200   
390                           Toyota Land Cruiser         54765   
194                                Jaguar XJR 4dr         74995   
330    Porsche 911 Carrera 4S coupe 2dr (convert)         84165   
198                          Jaguar XKR coupe 2dr         81995   
226                                  Lexus LX 470         64800   
269                        Mercedes-Benz S500 4dr         86970   
3                    Acura NSX coupe 2dr manual S         89765   
197                    Jaguar XKR convertible 2dr         86995   
166                                     Hummer H2         49995   
20                                  Audi RS 6 4dr         84600   
270           Mercedes-Benz SL500 convertible 2dr         90520   
258                       Mercedes-Benz CL500 2dr         94820   
217                    Land Rover Range Rover HSE         72250   
266                            Mercedes-Benz G500         76870   
271                    Mercedes-Benz SL55 AMG 2dr        121770   
272           Mercedes-Benz SL600 convertible 2dr        126670   
259              

Toyota Echo 2dr manual has the best price to MPG ratio and Porsche 911 GT2 2dr has the worst ratio.